### PR TITLE
Dengyi/feature/integrate ghw to goss

### PR DESCRIPTION
**Pull request by weakiwi**
##### Checklist

- [x] `make test-all` (UNIX) passes. CI will also test this
- [x] unit and/or integration tests are included (if applicable)
- [x] documentation is changed or added (if applicable)

### Description of change

Hello,
  This PR  wants to integrate ghw (https://github.com/jaypipes/ghw) function into goss. With ghw integration, goss will have more hardware level testing ability.

Thanks

### CODE REVIEW

**Comment by aelsabbahy** [comment]
Yeah, I think this will need to be rebased/submitted. The diff is difficult to read.

Looking forward to it, the new macid check is pretty cool

**Comment by weakiwi** [comment]
> Thank you @weakiwi.
> 
> I am having a hard time looking at your actual changes, so we will need to clean this up. Your code changes seen to have been committed with merges from the master branch so there is a lot of strange changes in this PR. Please rebase your changes to the master branch.
> 
> The PR is also tagged with `[draft]` is that still entended?

Hello @pedroMMM 
  

  Sorry for lately updateing my branch.
  

  The draft is because I want to integrate more feature from ghw, like:
      1. CPU (total cores, threading, model, etc)
      2. Block devices (type, size, partition, mountpoint, etc)
But I am not sure if we have any rules for importing 3rd part packages, so I do a minor changes for reviewers to check first. If the way I use 3rd part pkgs is ok, I will push other feature on the same branch.

THX

**Comment by weakiwi** [comment]
BTW. Shall I fix the issues in codeclimate. Seems those issues are not cause by my changes. THX

**Comment by pedroMMM** [comment]
> The draft is because I want to integrate more feature from ghw, like:
> 1. CPU (total cores, threading, model, etc)
> 2. Block devices (type, size, partition, mountpoint, etc)
> But I am not sure if we have any rules for importing 3rd part packages, so I do a minor changes for reviewers to check first. If the way I use 3rd part pkgs is ok, I will push other feature on the same branch.

@weakiwi That is a great approach. I just got confused about the messaging and didn't want to review a PR that wasn't ready nor leave you waiting for a review.

As far as I can tell there are no "rules" for importing 3rd party packages, besides the regular sanity checks. You can feel free to either add the new tests in this PR or in a new one, up to you. As it stands right now the PR is already adding value and functional so I would recommend merging it and just create new PRs for the new features since smaller PRs are much easier to manage.


> BTW. Shall I fix the issues in codeclimate. Seems those issues are not cause by my changes. THX

Since the issues aren't caused by your changes we aren't going to require you to address them. Usually, if you fixing issues its appreciated, but I took a look and these issues aren't very simple. So feel free to ignore them.

**Comment by weakiwi** [comment]
OK. Let me remove the `draft` tag and please help to merge.
I will add features on other small branches.
Thanks for your work!

**Comment by pedroMMM** [comment]
@aelsabbahy it's ready to merge when you have some time. 👍

**Comment by weakiwi** [comment]
Hello @aelsabbahy 
  
   Do you have any problem about this PR?

THX

**Comment by aelsabbahy** [comment]
I don't have a problem with this PR. Just waiting to merge it when I'm ready to cut a release. However, given the other PR that was submitted, I'm starting to wonder if I may not fully understand your goals with these PRs

I started an issue to discuss this feature and the other one here #570

File: system/interface.go [hunk 1]
```diff
@@ -41,13 +46,41 @@ func (i *DefInterface) setup() error {
...
+	ghwNetwork, err := ghw.Network()
+	if err != nil {
+		i.exists = false
+		i.err = err
+		return i.err
+	}
+	for _, nic := range ghwNetwork.NICs {
+		if nic.Name == i.name {
+			i.ghwIface = nic
+			break
```
Comment (aelsabbahy) [inline system/interface.go:58]: If this for any reason doesn't ever match. It causes a panic in the code later on when it tries to reference it.

**Pull request closed without merging**